File: wetty-chat-mobile/src/api/threads.ts
import type { AxiosResponse } from 'axios';
import type { MessagePreview, User } from './messages';
import apiClient from './client';

export type ThreadParticipant = User;

export interface ThreadListItem {
  chatId: string;
  chatName: string;
  chatAvatar: string | null;
  threadRootMessage: MessagePreview;
  participants: ThreadParticipant[];
  lastReply: MessagePreview | null;
  replyCount: number;
  lastReplyAt: string;
  unreadCount: number;
  lastReadMessageId: string | null;
  subscribedAt: string;
  archived: boolean;
}

/** Internal Redux state representation — replaces `lastReply` with a cache-only fallback. */
export interface StoredThreadListItem extends Omit<ThreadListItem, 'lastReply'> {
  cachedLastReply: MessagePreview | null;
}

export interface ListThreadsResponse {
  threads: ThreadListItem[];
  nextCursor: string | null;
}

export interface UnreadThreadCountResponse {
  unreadThreadCount: number;
  archivedUnreadThreadCount: number;
}

export interface ThreadSubscriptionStatusResponse {
  subscribed: boolean;
  archived: boolean;
}

export interface MarkThreadReadResponse {
  lastReadMessageId: string | null;
  unreadCount: number;
}

export function getThreads(params?: {
  limit?: number;
  before?: string;
  archived?: boolean;
}): Promise<AxiosResponse<ListThreadsResponse>> {
  const query: Record<string, string | number | boolean> = {};
  if (params?.limit != null) query.limit = params.limit;
  if (params?.before != null) query.before = params.before;
  if (params?.archived != null) query.archived = params.archived;
  return apiClient.get('/threads', { params: query });
}

export function markThreadAsRead(
  threadRootId: string,
  messageId: string,
): Promise<AxiosResponse<MarkThreadReadResponse>> {
  return apiClient.post(`/threads/${threadRootId}/read`, { messageId });
}

export function getUnreadThreadCount(): Promise<AxiosResponse<UnreadThreadCountResponse>> {
  return apiClient.get('/threads/unread');
}

export function subscribeToThread(
  chatId: string | number,
  threadRootId: string | number,
): Promise<AxiosResponse<void>> {
  return apiClient.put(`/chats/${chatId}/threads/${threadRootId}/subscribe`);
}

export function unsubscribeFromThread(
  chatId: string | number,
  threadRootId: string | number,
): Promise<AxiosResponse<void>> {
  return apiClient.delete(`/chats/${chatId}/threads/${threadRootId}/subscribe`);
}

export function getThreadSubscriptionStatus(
  chatId: string | number,
  threadRootId: string | number,
): Promise<AxiosResponse<ThreadSubscriptionStatusResponse>> {
  return apiClient.get(`/chats/${chatId}/threads/${threadRootId}/subscribe`);
}

export function archiveThread(chatId: string | number, threadRootId: string | number): Promise<AxiosResponse<void>> {
  return apiClient.put(`/chats/${chatId}/threads/${threadRootId}/archive`);
}

export function unarchiveThread(chatId: string | number, threadRootId: string | number): Promise<AxiosResponse<void>> {
  return apiClient.delete(`/chats/${chatId}/threads/${threadRootId}/archive`);
}

export interface ThreadReadStateResponse {
  lastReadMessageId: string | null;
}

export function getThreadReadState(threadRootId: string): Promise<AxiosResponse<ThreadReadStateResponse>> {
  return apiClient.get(`/threads/${threadRootId}/read-state`);
}

File: backend/src/dto/threads.rs
use chrono::{DateTime, Utc};
use serde::Serialize;

use crate::dto::{messages::MessagePreview, users::User};

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ThreadListItem {
    #[serde(with = "crate::serde_i64_string")]
    #[schema(value_type = String)]
    pub chat_id: i64,
    pub chat_name: String,
    pub chat_avatar: Option<String>,
    pub thread_root_message: MessagePreview,
    pub participants: Vec<User>,
    pub last_reply: Option<MessagePreview>,
    pub reply_count: i64,
    pub last_reply_at: DateTime<Utc>,
    pub unread_count: i64,
    #[serde(with = "crate::serde_i64_string::opt")]
    #[schema(value_type = Option<String>)]
    pub last_read_message_id: Option<i64>,
    pub subscribed_at: DateTime<Utc>,
    pub archived: bool,
}

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ListThreadsResponse {
    pub threads: Vec<ThreadListItem>,
    pub next_cursor: Option<String>,
}

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct MarkThreadReadResponse {
    #[serde(serialize_with = "crate::serde_i64_string::opt::serialize")]
    #[schema(value_type = Option<String>)]
    pub last_read_message_id: Option<i64>,
    pub unread_count: i64,
}

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct UnreadThreadCountResponse {
    pub unread_thread_count: i64,
    pub archived_unread_thread_count: i64,
    pub unread_message_count: i64,
    pub archived_unread_message_count: i64,
}

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ThreadSubscriptionStatusResponse {
    pub subscribed: bool,
    pub archived: bool,
}

#[derive(Serialize, utoipa::ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct ThreadReadStateResponse {
    #[serde(serialize_with = "crate::serde_i64_string::opt::serialize")]
    #[schema(value_type = Option<String>)]
    pub last_read_message_id: Option<i64>,
}
